File: part2/src/DistributedPuzzle_Actors/src/main/scala/GUI/PuzzleBoard.scala
package GUI

import com.fasterxml.jackson.module.scala.deser.overrides.MutableList
import model.PuzzleActor

import java.awt.event.{MouseAdapter, MouseEvent}
import java.awt.image.{BufferedImage, CropImageFilter, FilteredImageSource}
import java.awt.{BorderLayout, Color, GridLayout, Image}
import java.io.{File, IOException}
import java.util
import java.util.stream.IntStream
import java.util.{ArrayList, Collections}
import javax.imageio.ImageIO
import javax.swing.{BorderFactory, ImageIcon, JButton, JFrame, JOptionPane, JPanel, WindowConstants}
import scala.:+
import scala.collection.mutable.ListBuffer
import scala.language.postfixOps
import scala.util.Random

class PuzzleBoard(var rows: Int, var columns: Int, var imagePath: String, val actor: PuzzleActor) extends JFrame:
    
    var tiles: ListBuffer[Tile] = ListBuffer()
    val selectionManager: SelectionManager = SelectionManager(actor)

    setTitle("Puzzle")
    setResizable(false)
    setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE)

    val board: JPanel = JPanel()
    board.setBorder(BorderFactory.createLineBorder(Color.gray))
    board.setLayout(GridLayout(rows, columns, 0, 0))
    getContentPane.add(board, BorderLayout.CENTER)
    
    createTiles(imagePath)
    paintPuzzle(board)
    
    this.setVisible(true)

    def createTiles(imagePath: String): Unit =
        val image = ImageIO.read(File(imagePath))
        if (image == null)
            JOptionPane.showMessageDialog(this, "Could not load image", "Error", JOptionPane.ERROR_MESSAGE)
            return
        val w = image.getWidth(null)
        val h = image.getHeight(null)
        var p = 0
        val randomPositions: List[Int] = Random.shuffle(List.range(0, rows * columns))
        for (i <- 0 until rows; j <- 0 until columns)
            val imagePortion: Image = createImage(
                FilteredImageSource(
                    image.getSource,
                    CropImageFilter(
                        j * w / columns,
                        i * h / rows,
                        w / columns,
                        h / rows
                    )
                )
            )
            this.tiles = this.tiles :+ Tile(imagePortion, p, randomPositions(p))
            p = p + 1

    def paintPuzzle(board: JPanel): Unit =
        board.removeAll()
        this.tiles = this.tiles.sortWith(_.getCurrentPosition < _.getCurrentPosition)
        this.tiles.foreach(tile => {
            val button: JButton = JButton(new ImageIcon(tile.getImage))
            button.addActionListener(_ => {
                this.selectionManager.selectTile(tile, () => {
                    paintPuzzle(board)
                    checkSolution()
                }, true)
            })
            board.add(button)
        })
        pack()
        setLocationRelativeTo(null)
        
    def repaintPuzzle(): Unit =
        paintPuzzle(this.board)

    def checkSolution(): Unit =
        if (this.tiles.forall(tile => tile.isInRightPlace))
            JOptionPane.showMessageDialog(this, "Puzzle Completed!", "", JOptionPane.INFORMATION_MESSAGE)

File: part2/JavaRMI/src/main/java/org/example/puzzle/PuzzleBoard.java
package org.example.puzzle;

import javax.imageio.ImageIO;
import javax.swing.BorderFactory;
import javax.swing.JFrame;
import javax.swing.JOptionPane;
import javax.swing.JPanel;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.GridLayout;
import java.awt.Image;
import java.awt.image.BufferedImage;
import java.awt.image.CropImageFilter;
import java.awt.image.FilteredImageSource;
import java.io.File;
import java.io.IOException;
import java.io.Serializable;
import java.rmi.RemoteException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.IntStream;

public class PuzzleBoard extends JFrame implements Serializable {

    private final int rows;
    private final int columns;
    private final JPanel board;
	private List<Tile> tiles;
	private final SelectionManager selectionManager;

    public PuzzleBoard(
        final int rows,
        final int columns,
        final String imagePath,
        final List<Tile> tiles,
        final SelectionManager selectionManager
    ) {
    	this.rows = rows;
		this.columns = columns;
        this.tiles = tiles;
        this.selectionManager = selectionManager;
    	
    	setTitle("Puzzle");
        setResizable(false);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        
        board = new JPanel();
        board.setBorder(BorderFactory.createLineBorder(Color.gray));
        board.setLayout(new GridLayout(rows, columns, 0, 0));
        getContentPane().add(board, BorderLayout.CENTER);

        createTiles(imagePath);
        paintPuzzle();
    }

    
    private void createTiles(final String imagePath) {
		final BufferedImage image;
        try {
            image = ImageIO.read(new File(imagePath));
        } catch (IOException ex) {
            JOptionPane.showMessageDialog(this, "Could not load image", "Error", JOptionPane.ERROR_MESSAGE);
            return;
        }

        final int imageWidth = image.getWidth(null);
        final int imageHeight = image.getHeight(null);

        int position = 0;
        
        final List<Integer> randomPositions = new ArrayList<>();
        IntStream.range(0, rows * columns).forEach(randomPositions::add);
        Collections.shuffle(randomPositions);
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
            	final Image imagePortion = createImage(new FilteredImageSource(image.getSource(),
                    new CropImageFilter(
                        j * imageWidth / columns,
                        i * imageHeight / rows,
                        (imageWidth / columns),
                        imageHeight / rows
                    ))
                );
                tiles.add(new Tile(imagePortion, position, randomPositions.get(position)));
                position++;
            }
        }
	}
    
    private void paintPuzzle() {
    	board.removeAll();
    	
    	Collections.sort(tiles);
    	
    	tiles.forEach(tile -> {
    		final TileButton btn = new TileButton(tile);
            board.add(btn);
            btn.setBorder(BorderFactory.createLineBorder(Color.gray));
            btn.addActionListener(actionListener -> {
            	selectionManager.selectTile(tile, () -> {
            		paintPuzzle();
                	checkSolution();
            	});
            });
    	});
    	
    	pack();
        setLocationRelativeTo(null);
    }

    public void repaint() {
        paintPuzzle();
    }

    private void checkSolution() {
    	if (tiles.stream().allMatch(Tile::isInRightPlace)) {
    		JOptionPane.showMessageDialog(this, "Puzzle Completed!", "", JOptionPane.INFORMATION_MESSAGE);
    	}
    }

    public List<Tile> getTiles() {
        return tiles;
    }

    public void setTiles(List<Tile> tiles) {
        this.tiles = tiles;
    }



}
